### DISCUSSION TITLE
Can fastapi-cli be moved to an extra of fast api, rather than a default.

### DISCUSSION THREAD

**Discussion by ollz272** [fastapi/fastapi#11525] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
No code needed.
```

### Description

Hi,

First of all, thanks for your effort in maintaining fastapi, it is much appreciated.

The 0.111.0 release added a new feature, `fastapi-cli`, which also adds a number of new dependencies to fastapi by default. Whilst this feature looks useful, for us we would much rather not have these extra dependencies in our production environment, as we won't be using this feature when deployed.

Whilst there is the `fastapi-slim` package which offers fastapi without the cli, we dont feel this is an appropriate solution. We rely on other dependencies, such as [fastapi-restful](https://github.com/yuval9313/FastApi-RESTful/blob/master/pyproject.toml), which has a dependency on fastapi. Therefore even if we switch to use fastapi-slim on our project, the full fastapi will be installed because of this dependency. And since (poetry at least) doesn't support conditional dependencies, we can't switch this package over to use fastapi-slim without breaking backwards compatability with other fastapi versions. 

We believe the best solution here would be to have the cli be an extra. Given the cli is its own package anyway, having this be an extra (maybe under `cli`) would work just fine, and allow us to still use the regular fastapi packages, and those that dependencies on it, without needing to install the cli and all its dependencies.

We would therefore be greatful is the change making installing the CLI as a default be reverted and added as an opt-in extra.

Thanks

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.12.1

**Comment by zoola969** [comment]
What I got after upgrading fastapi:

httptools
markdown-it-py
mdurl
pygments
python-dotenv
pyyaml
rich
shellingham
uvloop
ujson
watchfiles
websockets

I think that installing a half of pypi is too much for useless feature, isn't it?

**Reply by ollz272** [reply]
We had the following extras installed:

dnspython (2.6.1)
email-validator (2.1.1)
Removing fastapi-cli (0.0.2)
Removing python-multipart (0.0.9)
Removing shellingham (1.5.4)
Removing typer (0.12.3)
ujson (5.9.0)

**Reply by tiangolo** [reply]
You can definitely get the minimum by installing `fastapi-slim` and then add anything extra you need on top. :nerd_face: 

---

Now, about the specific packages. For most of the extra ones you could get rid of them by installing `fastapi-slim`.

But there are some you might want to keep:

* httptools: this is used by Uvicorn to provide high-performance HTTP parsing, if you were not installing it, you should
* uvloop: also used by Uvicorn, this is the high-performance asyncio drop-in replacement, you should install it if you were not doing so

* rich: this is used by Typer and the CLI to provide the information display in the terminal. It will be used in the future by FastAPI to display nicer logs and tracebacks to improve the developer experience when fixing bugs in your own code using FastAPI. But again, that will also be optional (but probably useful).

* markdown-it-py: used by Rich IIRC
* mdurl: used by Rich IIRC
* pygments: used by Rich IIRC

* python-dotenv: You would probably use it to read config env vars, but you can remove it.
* pyyaml: this is part of `fastapi[all]`, not of `fastapi` or `fastapi-slim[standard]`, because it's used by the Starlette schema generator, but that's rarely used with FastAPI. Not sure why you would have it installed but you would probably want to remove it.

* shellingham: this package is tiny, it's used by Typer and the CLI to detect the terminal you are using, for shell completion. But you can remove it if you don't use that.

* ujson: this is used by the UJSONResponse, part of FastAPI, for high-performance responses (if you use it). But this will be much less important in a future version where Pydantic will handle the serialization directly on the Rust side.
* watchfiles: This is used by Uvicorn when you use `--reload` to detect code changes.
* websockets: Used to support web sockets.

---

* dnspython: Not sure if this is in the dependency tree for FastAPI :thinking: ...maybe it's needed by AnyIO
* email-validator: This is used by Pydantic for email validation, but if you don't use that, you can remove it.
* Removing fastapi-cli: you could instead install `fastapi-slim` (in the original case of the extra library on top, let's continue that conversation in my other answer asking about that)
* Removing python-multipart: this is needed if you want forms or file upload support, if you don't use it, you can remove it.
* Removing shellingham: shell detection for autocompletion, if you don't use it you can remove/discard it.
* Removing typer: used by FastAPI CLI, you can remove it/discard it if you don't use it.
* ujson: You could also remove this one, probably, it's only used with the UJSONResponse, you might not be using it.

**Reply by zoola969** [reply]
Thanks about revealing info about  httptools and uvloop. I missed it reading uvicorn docs. 
But rest of them are useless for me and I think that a python community should be more carefully with third-party libs. I'm quite upset with the fact that installing any lib I get a lot of strange dependencies, while half of them might be dead and other half might have vulnerabilities

**Comment by mjbommar** [comment]
Another reason: the `typer` dependency constraints conflict with common packages like `spacy`.

**Reply by tiangolo** [reply]
You are probably using a quite old version of spaCy if you are having conflicts with `typer`. In fact, one of the maintainers of spaCy is now also helping me maintain Typer. :sweat_smile: 

But anyway, you could get rid of it by installing `fastapi-slim`... just make sure you add the important extra dependencies for performance and all that (e.g. Uvicorn, httptools, uvloop, etc).

**Reply by mjbommar** [reply]
Unfortunately the version string constraints are conflicting on latest `spacy` (3.7.4) and `fastapi-cli` (0.0.2).  You can replicate like this to test:

```
$ poetry init
[...]
$ poetry add spacy fastapi-cli
Creating virtualenv y-Uxiwtdmj-py3.12 in /home/mjbommar/.cache/pypoetry/virtualenvs
Using version ^3.7.4 for spacy
Using version ^0.0.2 for fastapi-cli

Updating dependencies
Resolving dependencies... (0.2s)

Because no versions of fastapi-cli match >0.0.2,<0.0.3
 and fastapi-cli (0.0.2) depends on typer (>=0.12.3), fastapi-cli (>=0.0.2,<0.0.3) requires typer (>=0.12.3).
And because spacy (3.7.4) depends on typer (>=0.3.0,<0.10.0)
 and no versions of spacy match >3.7.4,<4.0.0, fastapi-cli (>=0.0.2,<0.0.3) is incompatible with spacy (>=3.7.4,<4.0.0).
So, because y depends on both spacy (^3.7.4) and fastapi-cli (^0.0.2), version solving failed.
```


* https://github.com/tiangolo/fastapi-cli/blob/main/pyproject.toml#L36
* https://github.com/explosion/spaCy/blob/master/requirements.txt#L12C1-L13C1

**Reply by aymenkrifa** [reply]
Installing FastAPI + SpaCy model (`en_core_web_md-3.7.1`) produces the following error:

> The conflict is caused by:
>     fastapi-cli 0.0.2 depends on typer>=0.12.3
>     spacy 3.7.4 depends on typer<0.10.0 and >=0.3.0
>     fastapi-cli 0.0.2 depends on typer>=0.12.3
>     spacy 3.7.2 depends on typer<0.10.0 and >=0.3.0

With the following dependencies:

- `fastapi==0.111.0`
- `spacy==3.7.4` (though spacy depends on a package called weasel that has typer as a dependency (typer>=0.3.0,<0.10.0)

**Comment by mattmess1221** [comment]
Why couldn't a new `cli` group be added for `fastapi-cli`? If I want just the cli instead of *everything*, I can just add `fastapi[cli]` instead of `fastapi-slim fastapi-cli-slim`

**Reply by tiangolo** [reply]
Although the change looks very simple, it has a bunch of nuances, in particular for newcomers, the developer experience gets complicated quickly when the simplest and default recommended packages are opt-in.

You can read more about it in my other response here: https://github.com/tiangolo/fastapi/discussions/11525#discussioncomment-9314953

**Comment by Chouvic** [comment]
We encountered same issue with version 0.111.0 when building fastapi from windows wheels. As the optional dependencies [standard] is instructed to be installed by default for both fastapi and fastapi-cli, this creates a Cyclic dependencies. I understand the fastapi is required for fastapi-cli, but not necessarily vice versa. Could we resolve this cyclic dependencies issue please?

Steps to reproduce:
pip install fastapi pipdeptree

$ pipdeptree --warn=fail
**Warning!! Cyclic dependencies found:**
* fastapi-cli => fastapi => fastapi-cli
* fastapi => fastapi-cli => fastapi
------------------------------------------------------------------------
httptools==0.6.1
pipdeptree==2.18.1
├── packaging [required: >=23.1, installed: 24.0]
└── pip [required: >=23.1.2, installed: 23.2.1]
python-dotenv==1.0.1
PyYAML==6.0.1
watchfiles==0.21.0
└── anyio [required: >=3.0.0, installed: 4.3.0]
    ├── idna [required: >=2.8, installed: 3.7]
    └── sniffio [required: >=1.1, installed: 1.3.1]
websockets==12.0

**Reply by tiangolo** [reply]
Oh, interesting, were you having issues installing and using FastAPI in Windows?

Or were the issues only unning that command `pipdeptree`?

Because it seems `pipdeptree` is complaining about a behavior that is officially supported by Python.

Cyclical dependencies are officially supported as long as they can be resolved (https://discuss.python.org/t/handling-installation-of-circular-dependencies/25531). :upside_down_face: (I checked that before defining the current setup).

**Reply by Chouvic** [reply]
No issue with installation but a warning from pipdeptree. Python community might allow it but Circular dependency smells, we should avoid it if we can.

**Comment by ns-ggeorgiev** [comment]
I don't think there's a need for slim or similar. Simply fastapi should not depend on fastapi-cli.

**Comment by tiangolo** [comment]
Thanks for the feedback!

So, the idea with having the new setup with `fastapi` apart from `fastapi-slim` is to make the **simplest developer experience** for the **majority of cases**, in particular for **newcomers**, as it's much easier to install `pip install fastapi` and have the default recommended packages.

---

For a newbie, `pip install fastapi[standard]` means explaining what the `[]` means, what is this `standard` keyword, what is the difference between `pip install fastapi` and `pip install fastapi[standard]`, and in many cases that command won't even work, because people might be using Zsh (e.g. in Mac by default) where it needs quotes around `pip install "fastapi[standard]"`, so, what should be the simplest use case for the newcomers becomes the most complex one, more difficult to explain and use.

Meanwhile, for any advanced developer (if you are here reading this you are advanced), if you want to trim some of the default extras, it should be a relatively simple process of installing `fastapi-slim` instead of `fastapi` and adding just the extra exact packages you need.

For example I imagine you would manually add `uvicorn`, and would make sure to install `uvicorn[standard]`, or independently install `httptools` and `uvloop` to get the performance benefits, etc.

**Reply by zoola969** [reply]
That's quite strange for me that you more care about newcomers than about current community and for instance making fastapi production ready(i mean major v1 version or supporting both pydantic version for smoother migration).

I understand that you want to push forward typer, but in my humble opinion cli is not the thing that is needed for web server. All developers that I know use either web servers(uvicorn, hypercorn, gunicorn etc) or just set up configuration in their IDEs.

**Reply by ollz272** [reply]
I completely disagree with the stance taken here and its clear to me at least that others feel the same.

**Reply by JP-Dhabolt** [reply]
@tiangolo Please rethink this stance, as it is not the expected behavior for a library.  Dependencies should ONLY include what is required for a module to function, not extra tooling, as this subverts expectations and results in unnecessary conflicts that makes it difficult to use the tool in a production environment.

The result of this change means that I have to pin `fastapi` to <0.111.0, as other modules have `fastapi` (not `fastapi-slim`) defined as a dependency, and there are a number of dependencies now included by default that cause bloat and conflict with other dependencies I rely on.

There should instead be a "meta" module that pulls in all of these extra dependencies called `fastapi-complete` if the desire is to have a single module install without having to explain "extras" to people.  The existing module of `fastapi` should not be converted into a meta module.

I appreciate that you are trying to solve a problem that you have as an open source maintainer (a thankless job, which I appreciate tremendously), but I would ask that you please consider publishing the meta package as the new package, and leave `fastapi` as the standard (and beloved) library without bloat.

**Reply by hasansezertasan** [reply]
> For a newbie, `pip install fastapi[standard]` means explaining what the `[]` means, what is this `standard` keyword, what is the difference between `pip install fastapi` and `pip install fastapi[standard]`, and in many cases that command won't even work, because people might be using Zsh (e.g. in Mac by default) where it needs quotes around `pip install "fastapi[standard]"`, so, what should be the simplest use case for the newcomers becomes the most complex one, more difficult to explain and use.

You are the person who taught me python type annotations through FastAPI Documentation.

You are really good at documenting.

I think the best person to teach (give a tutorial about) optional dependencies is you.

It's not hard to explain `pip install fastapi[cli]` in the fastapi cli documentation. Let me try:

> To use FastAPI's CLI, you should install extra dependencies using `pip install fastapi[cli]` command.

All alone, I'm with @JP-Dhabolt.

**Reply by mattmess1221** [reply]
@hasansezertasan Completely agree. And if quoting issues with zsh users on mac are that much of an issue, recommending use of requirements.txt is both a workaround and good practice.

**Reply by benjsec** [reply]
Just adding my voice here, are we any closer to getting this issue resolved?

**Reply by aybruhm** [reply]
> Just adding my voice here, are we any closer to getting this issue resolved?

Unfortunately, the issue was marked as completed when tiangolo marked his comment as the solution.

**Reply by RobertDeRose** [reply]
I hate to pile on, but this being considered the solution is sadly not a great resolution.
While I appreciate that FastAPI is still sub v1 status, many people rely on it as it stands. While this is not a breaking change, it does now force the addition of many new packages that were not previously needed. This adds bloat as others have stated and if you are working in a limited environment, this can cause harm.

The goal of any open source project, especially when it's this popular and this big, is to first do no harm with new features.

It also does not work out of the box for all setups. I for one use a Factory pattern for creating my `app` because it allows some flexibility I need and like to have to use the same code for my API as well as my CLI.

**Comment by tiangolo** [comment]
Apart from the reason to have a `fastapi` and `fastapi-slim` (in another response here), let's talk about the specific use case of the original post with https://github.com/yuval9313/FastApi-RESTful @ollz272 

The first question I have would be, why do you need that package instead of using FastAPI directly?

I imagine it's because of some of the features/characteristics, but I think there's an official way to achieve each one with pure FastAPI... or maybe I'm missing the point in some of those.

Which specific features do you need from that package that you can't do with FastAPI, and why are those features useful?

**Reply by zoola969** [reply]
I suppose that some people need such all-inclusive frameworks because they were using django before. But I thought that purpose of starlette-fastapi was to be minimalistic. It's great that if you want a simple web app you just install starlette, and if you want to get REST api you need to just install fastapi. I believe that django is losing popularity because of it's gigantism. In some point it became so large that it lost ability to develop fast.

I like flask style: you have a minimal core functionality and if you need more - just install third party extensions or write whatever you need

**Comment by mattmess1221** [comment]
Another seemingly unintended consequence of this: corruption of the virtualenv when switching between fastapi and fastapi-slim via a git checkout or stash. It's caused because both `fastapi` and `fastapi-slim` ship the same packages, so installing both then uninstalling one will remove all the files, but package tools will still see it as installed.

Recreation:

```
pdm add fastapi
git commit -A
pdm remove fastapi
pdm add fastapi-slim fastapi-cli-slim
git stash
pdm sync --clean <-- suspect, but still works. uninstalls fastapi-slim, then installs fastapi
git stash pop
pdm sync --clean <-- broken, installs fastapi-slim, then uninstalls fastapi
ls -d .venv/lib/python3.12/site-packages/fastapi # no such file or directory
```

Second PDM Output:
```
 % pdm sync --clean
Synchronizing working set with resolved packages: 8 to add, 0 to update, 1 to remove

  ✔ Install fastapi-cli-slim 0.0.2 successful
  ✔ Install mdurl 0.1.2 successful
  ✔ Install shellingham 1.5.4 successful
  ✔ Install typer 0.12.3 successful
  ✔ Install fastapi-slim 0.111.0 successful
  ✔ Remove fastapi 0.110.2 successful
  ✔ Install markdown-it-py 3.0.0 successful
  ✔ Install rich 13.7.1 successful
  ✔ Install pygments 2.18.0 successful

🎉 All complete!


% ls -d .venv/lib/python3.12/site-packages/fastapi 
".venv/lib/python3.12/site-packages/fastapi": No such file or directory (os error 2)
```

**Comment by hwong557** [comment]
Upon upgrading FastAPI to 0.111.0, we started randomly getting errors:
```
OSError: [Errno 9] Bad file descriptor 
```

It turns out this is caused by `uvloop`. Please see:

* https://github.com/benoitc/gunicorn/issues/1877
* https://github.com/MagicStack/uvloop/issues/414

So the default inclusion of `uvloop` may not be a great choice. It seems appear in k8s but not locally, which may confuse beginners.

Thank you very much for your work in developing this very fantastic project.

**Comment by tiangolo** [comment, marked as answer]
Thanks for the feedback everyone!

As having `fastapi` include by default the standard dependencies has been inconvenient to several people, I decided to make those standard dependencies an optional extra group `standard`.

So, to get the standard optional dependencies people now have to install it with:

```bash
pip install "fastapi[standard]"
```

**Note**: remember to use the quotes, depending on your terminal that might be required.

Having the optional extra, square brackets, quotes, is not as nice as just `fastapi`, it's extra things to have in mind, and extra things for newcomers to learn from the beginning. But anyway, it seems that having the default standard dependencies installed by default, and having to change the installed package to be `fastapi-slim` when people didn't want those standard dependencies instead was being too inconvenient for some people so I decided to switch back to have `fastapi` as the minimal/slim version and the standard extra dependencies in the group `standard`.

This is available in FastAPI version 0.112.0. :tada:

**Reply by MtkN1** [reply]
Thank you for making decisions that consider community feedback! Having minimal and appropriate dependencies is crucial for ensuring usability in production environments.

By the way, what will happen to the `fastapi-slim` package now that it has served its purpose? Currently, the same version `0.112.0` is published on PyPI for both `fastapi` and `fastapi-slim`. Will it be maintained for compatibility purposes?
https://pypi.org/project/fastapi-slim/#history

**Reply by tiangolo** [reply]
Thanks @MtkN1! Yes, `fastapi-slim` will be a mirror of `fastapi` for compatibility purposes, at least for a while.
